datacube/index/memory/_datasets.py

```diff
@@ -598,36 +598,36 @@ def search_returning(self,
                          archived: bool | None = False,
                          order_by: str | Field | None = None,
                          **query: QueryField) -> Iterable[Tuple]:
+        # Note that this implementation relies on dictionaries being ordered by insertion - this has been the case
+        # since Py3.6, and officially guaranteed behaviour since Py3.7.
         if order_by:
             raise ValueError("order_by argument is not currently supported by the memory index driver.")
         if field_names is None and custom_offsets is None:
-            field_names = self._index.products.get_field_names()
+            field_name_d = {f: None for f in self._index.products.get_field_names()}
         elif field_names:
-            field_names = list(field_names)
+            field_name_d = {f: None for f in field_names}
         else:
-            field_names = []
-        field_name_set = set(field_names)
+            field_name_d = {}
 
         if custom_offsets:
             custom_fields = {
                 name: SimpleField(offset, lambda x: x, "any", name=name, description="")
                 for name, offset in custom_offsets.items()
             }
             for name in custom_fields:
-                if name not in field_name_set:
-                    field_name_set.add(name)
-                    field_names.append(name)
+                if name not in field_name_d:
```

Comment:
I don't think you even need the check since there's no risk of duplicates with a dict